evcache-core/src/main/java/com/netflix/evcache/EVCacheImpl.java

```diff
@@ -512,12 +533,167 @@ <T> T doGet(EVCacheKey evcKey , Transcoder<T> tc) throws EVCacheException {
-            if (log.isDebugEnabled() && shouldLog()) log.debug("GET : APP " + _appName + ", Took " + duration + " milliSec.");
+            if (log.isDebugEnabled() && shouldLog())
+                log.debug("GET : APP " + _appName + ", Took " + duration + " milliSec.");
+        }
+    }
+
+    public <T> CompletableFuture<T> get(String key, Transcoder<T> tc, ExecutorService executorService) throws EVCacheException {
+        if (null == key) throw new IllegalArgumentException("Key cannot be null");
+        final EVCacheKey evcKey = getEVCacheKey(key);
+        return doAsyncGet(evcKey, tc, executorService);
+    }
+
+    <T> CompletableFuture<T> doAsyncGet(EVCacheKey evcKey, Transcoder<T> tc, ExecutorService executorService) throws EVCacheException {
+        final boolean throwExc = doThrowException();
+        EVCacheClient client = buildEvCacheClient(throwExc);
+        if (client == null) {
+            return null;
+        }
+        final EVCacheEvent event = buildAndStartEvent(client, evcKey, throwExc);
+        final long start = EVCacheMetricsFactory.getInstance().getRegistry().clock().wallTime();
+        StringBuilder status = new StringBuilder(EVCacheMetricsFactory.SUCCESS);
+        StringBuilder cacheOperation = new StringBuilder(EVCacheMetricsFactory.YES);
+        AtomicInteger tries = new AtomicInteger(1);
+        final boolean hasZF = hasZoneFallback();
+        boolean throwEx = !hasZF && throwExc;
+        return getAsyncData(client, evcKey, tc, throwEx, hasZF, executorService)
+                .thenCompose(data -> handleRetry(data, evcKey, tc, client, hasZF, throwExc, event, executorService))
+                .thenApply(data -> handleData(data, event, evcKey, client, cacheOperation))
+                .exceptionally(ex -> handleException(ex, throwExc, event, evcKey))
+                .handle((data, ex) -> handleFinally(data, status, tries, client, cacheOperation, start));
+    }
+
+    private EVCacheClient buildEvCacheClient(boolean throwExc) throws EVCacheException {
+        EVCacheClient client = _pool.getEVCacheClientForRead();
+        if (client == null) {
+            incrementFastFail(EVCacheMetricsFactory.NULL_CLIENT, Call.COMPLETABLE_FUTURE_GET);
+            if (throwExc) throw new EVCacheException("Could not find a client to get the data APP " + _appName);
+            return null;
+        }
+        return client;
+    }
+
+    private EVCacheEvent buildAndStartEvent(EVCacheClient client, EVCacheKey evcKey, boolean throwExc) throws EVCacheException {
+        EVCacheEvent event = createEVCacheEvent(Collections.singletonList(client), Call.COMPLETABLE_FUTURE_GET);
+        if (event != null) {
+            event.setEVCacheKeys(Arrays.asList(evcKey));
+            try {
+                if (shouldThrottle(event)) {
+                    incrementFastFail(EVCacheMetricsFactory.THROTTLED, Call.COMPLETABLE_FUTURE_GET);
```

Comment:
similar comment as buildEvCacheClient method. This method name seems generic, but it increments metric against COMPLETABLE_FUTURE_GET